### PR TITLE
create custom html syntax for markdown

### LINKED ISSUE #454: [content] Markdown children within HTML blocks are not included as children
Markdown children within HTML blocks (in a Markdown file) are rendered outside the parent HTML element.

For example:

```md
<details>
<summary>A summary of my details element.</summary>

The paragraph within my details element.

</details>
```

Gets rendered as:

```html
<details>
<summary>A summary of my details element.</summary>
</details>
<p>The paragraph within my details element.</p>
```

However, I expect it to render as:

```html
<details>
<summary>A summary of my details element.</summary>
<p>The paragraph within my details element.</p>
</details>
```

https://dart-lang.github.io/tools/markdown/ handles this case fine, so it seems it might be an issue introduced in the provided `MarkdownParser`.

---

\cc @schultek This is the largest blocker I'm facing now. If you have any pointers, I'll hopefully have some time to explore a fix soon.

### PR DESCRIPTION
**Pull request by schultek**
## Description

Fixes #454

## Type of Change

- 🛠️ Bug fix

## Ready Checklist

- [ ] I've read the [Contribution Guide](https://github.com/schultek/jaspr/blob/main/CONTRIBUTING.md).
- [ ] In case this PR changes one of the core packages, I've modified the respective **CHANGELOG.md** file using
      the [semantic_changelog](https://github.com/rrousselGit/semantic_changelog) format.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added myself to the AUTHORS file (optional, if you want to).

### CODE REVIEW

**Comment by schultek** [comment]
@parlough This should fix the html parsing issue. Can you help me test this by running this branch on your project?

**Comment by parlough** [comment]
> @parlough This should fix the html parsing issue. Can you help me test this by running this branch on your project?

Thanks for investigating and working on this @schultek!

The general rendering output is much better now. However, it seems the syntax breaks when using a [void element](https://developer.mozilla.org/en-US/docs/Glossary/Void_element) such as `` or `<br>`. Everything after one of those elements usages is missing from the output.

For example, in the following Markdown:

```md
**Hello** world!

<img src="/assets/the-world.png" alt="A depiction of parlough waving at the world.">

**Goodbye** world!
```

"Goodbye world" isn't being rendered.

**Comment by schultek** [comment]
It should work with proper self-closing tags like ``.

Detecting the ones you showed could be difficult. Except for what I could do is hardcode certain tags like img and br to always selfclose

**Comment by parlough** [comment]
> Except for what I could do is hardcode certain tags like img and br to always selfclose.

I think that's fine to do and I guess it's what other HTML parsers do. The HTML standard outlines a very small list of "Void elements" that can't have content (https://html.spec.whatwg.org/multipage/syntax.html#void-elements) and have this behavior:

- `area`
- `base`
- `br`
- `col`
- `embed`
- `hr`
- `img`
- `input`
- `link`
- `meta`
- `source`
- `track`
- `wbr`

**Comment by schultek** [comment]
I remembered I already have this list here: https://github.com/schultek/jaspr/blob/main/packages%2Fjaspr%2Flib%2Fsrc%2Fserver%2Fmarkup_render_object.dart#L173

So I can reuse it

**Comment by schultek** [comment]
@parlough Should be fixed now, can you try again.

**Comment by parlough** [comment]
> @parlough Should be fixed now, can you try again.

I have only had a chance to test a little bit and it seems fixed when the void element is at the root level of the Markdown, but  it seems there still might be an issue once they're more deeply nested.

For example:

```md
<a href="/get-started" class="large-button">
  <img src="/assets/the-world.png" alt="A depiction of parlough waving at the world.">
  <span>Hello world!</span>
</a>
```

Seems to cause the `span` to be missing, rendering as:

```html
<a href="/get-started" class="large-button">
  <img src="/assets/the-world.png" alt="A depiction of parlough waving at the world.">
</a>
```

**Comment by parlough** [comment]
Another _(but much less important)_ issue that I'm hitting:

---

I'm also having some issues with [processing instructions](https://developer.mozilla.org/en-US/docs/Web/API/ProcessingInstruction) left in the Markdown.

For example, in a Markdown file:

````markdown
<?some-instruction ?>
```dart
void main() {
  print('Hello');
}
```
````

Results in:

```text
[SERVER] [ERROR] Assertion failed: "Failed to parse HTML block: ```dart. Error: Assertion failed: \"Unexpected parse error: expected-tag-name-but-got-question-mark\""                                                                                                                  
[SERVER] [ERROR] package:jaspr_content/src/page_parser/markdown_parser.dart 114:9  CustomHtmlSyntax.parse
[SERVER] [ERROR] package:markdown/src/block_parser.dart 178:32                     BlockParser.parseLines
[SERVER] [ERROR] package:markdown/src/document.dart 87:44                          Document.parseLineList
[SERVER] [ERROR] package:markdown/src/document.dart 83:36                          Document.parse
[SERVER] [ERROR] package:parlough_site/markdown/markdown_parser.dart 30:45         ParloughMarkdownParser.parsePage
[SERVER] [ERROR] package:jaspr_content/src/page_parser/page_parser.dart 32:19      PageBuilderExtension.parsePage
[SERVER] [ERROR] package:jaspr_content/src/page.dart 296:27                        PageHandlersExtension.parseNodes
[SERVER] [ERROR] package:jaspr_content/src/page.dart 125:17                        Page.build
[SERVER] [ERROR] package:jaspr_content/src/page.dart 120:19                        Page.render.<fn>
```

---


The MDN page [suggests](https://developer.mozilla.org/en-US/docs/Web/API/ProcessingInstruction#:~:text=In%20these%2C%20a%20process%20instruction%20will%20be%20considered%20as%20a%20comment%20and%20be%20represented%20as%20a%20Comment%20object%20in%20the%20tree.) HTML documents can ignore them and parse them as comment nodes. Do you think that'd be possible?

**Comment by parlough** [comment]
I've opened an _(extremely)_ WIP PR with some of my exploration in https://github.com/dart-lang/site-www/pull/6702. Perhaps some of the pages can prove as helpful debugging examples.

Tomorrow I'll work on identifying pages that still have problems and create a list of them with what the problem is, but I thought I'd post the source for now in case it's helpful.

Thanks again 💙

**Comment by schultek** [comment]
@parlough 

> I have only had a chance to test a little bit and it seems fixed when the void element is at the root level of the Markdown, but it seems there still might be an issue once they're more deeply nested.

Fixed, I missed one more check.

> I'm also having some issues with [processing instructions](https://developer.mozilla.org/en-US/docs/Web/API/ProcessingInstruction) left in the Markdown.

Done, these are now parsed to comments. 

```html
<?some-instruction ?>
```

will result in the final html as:

```html
<!--some-instruction -->
```

File: packages/jaspr_content/lib/src/page_parser/markdown_parser.dart [hunk 1]
```diff
@@ -44,145 +47,163 @@ class MarkdownParser implements PageParser {
     final nodes = <Node>[];
     for (final node in markdownNodes) {
       if (node is md.Text) {
-        nodes.addAll(HtmlParser.buildNodes(html.parseFragment(node.text).nodes));
+        nodes.add(TextNode(node.text));
```
Comment (parlough) [inline packages/jaspr_content/lib/src/page_parser/markdown_parser.dart]: HTML blocks seem to be working well now, thanks!

However it seems inline HTML such as `<sub></sub>` is rendering now as text and content like `&&` is rendering as its escaped equivalent, even within Markdown backticks. Reverting this line change seems to fix at least the few cases I checked. Is there too much escaping happening somewhere?

Comment (schultek) [reply packages/jaspr_content/lib/src/page_parser/markdown_parser.dart]: I should add some tests 👀

Ideally all parsing should be done by the html syntax so I would aim to keep this extra parsing removed

Comment (schultek) [reply packages/jaspr_content/lib/src/page_parser/markdown_parser.dart]: Done, tests added.

I needed to revert that to correctly support inline html. To not have that I would need a custom inline syntax as well, but thats too much effort for now.

**Comment by parlough** [comment]
After checking a bunch of pages, the only usage I'm still having issues with is with Markdown code blocks and inline code nested in HTML. Specifically when they include generics, as they're getting interpreted as HTML tags.

````markdown
<div>

```dart
void method<T>(T t) {
  print(t);
}
```

</div>
````

**Comment by schultek** [comment]
> After checking a bunch of pages, the only usage I'm still having issues with is with Markdown code blocks and inline code nested in HTML. Specifically when they include generics, as they're getting interpreted as HTML tags.

Oh no.

I think I need to rethink my approach here, this is a major problem.

**Comment by schultek** [comment]
Ok. I changed the implementation (again), this should hopefully do it. All tests pass.

Its also a lot cleaner now and using the standard html block syntax as a foundation, meaning the rules for handling html are exactly the same as in the spec. For example to put markdown content inside html tags requires an empty line between the start tag and the content.

> I think consider landing this as it's quite large, but don't make the releases yet as we continue to land some other improvements and breaking changes. It'll also give some time for me to identify any other edge cases.

Yes agreed, I'm happy to land this after you give it one last try. :)

**Comment by parlough** [comment]
> Yes agreed, I'm happy to land this after you give it one last try. :)

I didn't test everything but the previous cases I had trouble with seem to be working as well as a few additional pages that weren't working at all before. Thanks! Plus this new implementation direction seems a lot easier to maintain :)
